integrations/pjrt/src/iree_pjrt/common/api_impl.cc

```diff
@@ -852,6 +914,127 @@ iree_status_t DeviceInstance::HostBufferToDeviceZeroDim(
   return iree_ok_status();
 }
 
+iree_status_t DeviceInstance::TransposeBroadcastDeviceBuffer(
+    BufferInstance* buffer, iree_hal_element_type_t element_type,
+    const iree_hal_dim_t* input_dims, const iree_hal_dim_t* output_dims,
+    const int64_t* perms, size_t num_dims,
+    PJRT_HostBufferSemantics host_buffer_semantics,
+    EventInstance** out_done_with_host_buffer_event,
+    BufferInstance** out_buffer) {
+  if (num_dims > 9) {
+    auto ret = iree_make_status(IREE_STATUS_INVALID_ARGUMENT,
+                                "number of dimensions exceeded 9");
+  }
+
+  std::array<iree_hal_dim_t, 9> transpose_dims;
+  for (int i = 0; i < num_dims; ++i) {
+    transpose_dims[i] = input_dims[perms[i]];
+  }
+
+  auto typeBuilder = [](const iree_hal_dim_t* dims, int64_t num_dims,
+                        const char* ty) {
+    std::stringstream ss;
+    ss << "tensor<";
+    for (int i = 0; i < num_dims; ++i) {
+      ss << dims[i] << "x";
+    }
+
+    ss << ty << ">";
+    return ss.str();
+  };
+
+  auto arrayBuilder = [](const int64_t* vals, int64_t sz) {
+    std::stringstream ss;
+    ss << " {permutation = dense<[" << vals[0];
+    for (int i = 1; i < sz; ++i) ss << ", " << vals[i];
+    ss << "]> : tensor<" << sz << "xi64>}";
+    return ss.str();
+  };
+
+  auto broadcastBuilder = [](int64_t sz) {
+    std::stringstream ss;
+    ss << "{broadcast_dimensions = dense<[0";
+    for (int i = 1; i < sz; ++i) ss << ", " << i;
+    ss << "]> : tensor<" << sz << "xi64>}";
+    return ss.str();
+  };
+
+  const char* mlir_ty;
+  IREE_RETURN_IF_ERROR(
+      PJRTApiConverter::MapElementTypeToMlirType(element_type, &mlir_ty));
+
+  auto input_ty = typeBuilder(input_dims, num_dims, mlir_ty);
+  auto transpose_ty = typeBuilder(transpose_dims.data(), num_dims, mlir_ty);
+  auto output_ty = typeBuilder(output_dims, num_dims, mlir_ty);
+  auto perms_str = arrayBuilder(perms, num_dims);
+  auto broadcast_str = broadcastBuilder(num_dims);
+
+  std::stringstream program_builder;
```

Comment:
Could we use formatv here? May be simpler to read and update

Reply:
(I mean like a `R"(` string kind)

Reply:
Managed to rebuild with snprintf. Should be cleaner, I just wish C++ had the builtin formatting methods python has.